lib/os/onoff.c

```diff
@@ -650,3 +661,255 @@ int onoff_sync_finalize(struct onoff_sync_service *srv,
+{
+	sys_snode_t *new_head = sortlist_head_peek(&mgr->future_clients);
+
+	if (head == new_head) {
+		return;
+	}
+	/* closest request changed. */
+	if (head) {
+		z_abort_timeout(&mgr->timer);
+	}
+
+	if (new_head) {
+		k_timeout_t t = get_delayed_cli(new_head)->timeout;
+
+		if (noffset) {
+			t = Z_TIMEOUT_TICKS(
+				Z_TICK_ABS(Z_TICK_ABS(t.ticks) - noffset));
+		}
+
+		z_add_timeout(&mgr->timer, timeout_handler, t);
+	}
+}
+
+static inline bool is_on_or_to_on(uint32_t flags)
+{
+	return (flags & ONOFF_STATE_ON) == ONOFF_STATE_ON;
+}
+
+static inline uint32_t get_noffset(struct onoff_delayed_manager *mgr)
+{
+	return is_on_or_to_on(mgr->mgr.flags) ? mgr->stop_time : 0;
+}
+
+static void timeout_handler(struct _timeout *t)
+{
+	struct onoff_delayed_client *cli;
+	int rv;
+	struct onoff_delayed_manager *mgr =
+		CONTAINER_OF(t, struct onoff_delayed_manager, timer);
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&mgr->mgr.lock);
+	sys_snode_t *head = sortlist_get(&mgr->future_clients);
+
+	reset_timer(mgr, head, get_noffset(mgr));
+
+	cli = get_delayed_cli(head);
+	rv = request_and_unlock(&mgr->mgr, &cli->cli, key);
+	if (rv < 0) {
+		onoff_client_callback cb =
+		(onoff_client_callback)sys_notify_finalize(&cli->cli.notify,
+								rv);
+
+		if (cb) {
+			cb(&mgr->mgr, &cli->cli, ONOFF_STATE_ERROR, rv);
+		}
+	}
+}
+
+int onoff_delayed_request(struct onoff_delayed_manager *mgr,
+			  struct onoff_delayed_client *cli,
+			  k_timeout_t delay)
+{
+	int rv = validate_args(&mgr->mgr, &cli->cli);
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	/* Convert to exact timeout. If time behind trigger request. */
+	if (get_abs_time(&cli->timeout, delay, mgr->startup_time)) {
+		return onoff_request(&mgr->mgr, &cli->cli);
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&mgr->mgr.lock);
+	sys_snode_t *head = sortlist_head_peek(&mgr->future_clients);
+
+	sortlist_add(&mgr->future_clients, &cli->cli.node);
+	reset_timer(mgr, head, get_noffset(mgr));
+
+	k_spin_unlock(&mgr->mgr.lock, key);
+
+	return rv;
+}
+
+int onoff_delayed_cancel(struct onoff_delayed_manager *mgr,
+			 struct onoff_delayed_client *cli)
+{
+	k_spinlock_key_t key = k_spin_lock(&mgr->mgr.lock);
+	sys_snode_t *head = sortlist_head_peek(&mgr->future_clients);
+
+	if (sortlist_remove(&mgr->future_clients, &cli->cli.node)) {
+		reset_timer(mgr, head, get_noffset(mgr));
+		k_spin_unlock(&mgr->mgr.lock, key);
+		return 0;
+	}
+
+	k_spin_unlock(&mgr->mgr.lock, key);
+
+	return onoff_cancel(&mgr->mgr, &cli->cli);
+}
+
+static void onoff_delayed_monitor(struct onoff_manager *mgr,
+				  struct onoff_monitor *mon,
+				  uint32_t state,
+				  int res)
+{
+	if (state == ONOFF_STATE_TO_ON || state == ONOFF_STATE_TO_OFF) {
+		struct onoff_delayed_manager *dmgr =
+			CONTAINER_OF(mgr, struct onoff_delayed_manager, mgr);
+		k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+		sys_snode_t *head = sortlist_head_peek(&dmgr->future_clients);
+
+		if (head) {
+			uint32_t offset = state == ONOFF_STATE_TO_ON ?
```

Comment:
OK, I see how the onoff monitor is used for one of the conditions I identified.

What a monitor-based solution doesn't do is handle the case I identified where to meet the deadline a transition in the on-off service must be inhibited.

Also I'm not entirely convinced that looking only at the head of the clients queue when sorted by adjusted deadline is sufficient, if the start and stop delays are different.  That's partly why I suggest storing the actual deadline: it makes it much easier to see when a state has to be reached, not when something has to be done to reach a state at a deadline that's no longer stored.

And this still seems to care only about transitions to on, while transitions to off are equally important for some use cases.